### PR TITLE
Component array enzyme fix

### PR DESCRIPTION
**Pull request by benjaminfaber**
The current implementation with `ComponentArrays` and `Enzyme` does not work, the fix is relatively simple. An explicit type conversion to the input data is required.  This has been implemented with `X(dkx)`, as `convert(typeof(X), dₖx)` was not able to find the correct conversion type.  Is this the desired way to fix this issue? Or will I need to first implement a fix in `ComponentArrays` to handle `convert(typeof(X), dₖx)` properly?

I also added the similar method, but for just a `Duplicated` input argument, is this appropriate? I understand using `BatchDuplicated` to compute a Jacobian, but won't `Duplicated` allow for more efficient gradient calculations?

### CODE REVIEW

**Comment by gdalle** [comment]
Hi @benjaminfaber, thanks for your contribution!
Can you come up with an MWE of what fails and how? This seems like an easy fix indeed but I'd like to investigate where the conversion becomes necessary for Enzyme.
As for `Duplicated`, I'm not an expert in writing Enzyme rules but I don't think it should make too much of a difference? In any case, what we really need is a reverse Enzyme rule (#35) but I haven't gotten around to it yet.

**Comment by benjaminfaber** [comment]
Yes, here's a MWE using the ComponentArrays tutorial from the docs:
```julia
using ImplicitDifferentiation
using Enzyme
using ComponentArrays

function forward_components_aux(a::AbstractVector, b::AbstractVector, m::Number)
    d = m * sqrt.(a)
    e = sqrt.(b)
    return d, e
end

function conditions_components_aux(a, b, m, d, e)
    c_d = (d ./ m) .^ 2 .- a
    c_e = (e .^ 2) .- b
    return c_d, c_e
end;

function forward_components(x::ComponentVector)
    d, e = forward_components_aux(x.a, x.b, x.m)
    y = ComponentVector(; d=d, e=e)
    return y
end

function conditions_components(x::ComponentVector, y::ComponentVector)
    c_d, c_e = conditions_components_aux(x.a, x.b, x.m, y.d, y.e)
    c = ComponentVector(; c_d=c_d, c_e=c_e)
    return c
end;

implicit_components = ImplicitFunction(forward_components, conditions_components);

a, b, m = [1.0, 2.0], [3.0, 4.0, 5.0], 6.0
x = ComponentVector(; a=a, b=b, m=m)
dx_zero = Enzyme.make_zero(x)
dx = Vector{typeof(x)}(undef, length(x))

for i in eachindex(dx)
    fill!(dx_zero, 0.)
    dx_zero[i] = 1.
    dx[i] = copy(dx_zero)
end
dx = tuple(dx...)

Enzyme.autodiff(Enzyme.Forward, implicit_components, Enzyme.BatchDuplicated, Enzyme.BatchDuplicated(x, dx))
```
This gives the following error stack:
```julia
julia> show(err)
1-element ExceptionStack:
MethodError: no method matching Duplicated(::ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, ::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}})

Closest candidates are:
  Duplicated(::T1, ::T1) where T1
   @ EnzymeCore ~/.julia/packages/EnzymeCore/yDAqg/src/EnzymeCore.jl:66
  Duplicated(::T1, ::T1, ::Bool) where T1
   @ EnzymeCore ~/.julia/packages/EnzymeCore/yDAqg/src/EnzymeCore.jl:66

Stacktrace:
  [1] pushforward(f::ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, backend::ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, x::ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, dx::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, ::DifferentiationInterface.NoPushforwardExtras)
    @ DifferentiationInterfaceEnzymeExt ~/.julia/packages/DifferentiationInterface/6pnz2/ext/DifferentiationInterfaceEnzymeExt/forward_onearg.jl:24
  [2] pushforward!(f::ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, dy::Vector{Float64}, backend::ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, x::ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, dx::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, extras::DifferentiationInterface.NoPushforwardExtras)
    @ DifferentiationInterfaceEnzymeExt ~/.julia/packages/DifferentiationInterface/6pnz2/ext/DifferentiationInterfaceEnzymeExt/forward_onearg.jl:55
  [3] (::ImplicitDifferentiation.PushforwardOperator!{ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, DifferentiationInterface.NoPushforwardExtras, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}})(res::Vector{Float64}, v::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, α::Float64, β::Float64)
    @ ImplicitDifferentiation ~/.julia/dev/ImplicitDifferentiation/src/operators.jl:84
  [4] mul!(res::Vector{Float64}, op::LinearOperators.LinearOperator{Float64, Int64, ImplicitDifferentiation.PushforwardOperator!{ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, DifferentiationInterface.NoPushforwardExtras, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}}, DataType, Nothing, Vector{Float64}}, v::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, α::Float64, β::Float64)
    @ LinearOperators ~/.julia/packages/LinearOperators/LoNOe/src/operations.jl:29
  [5] mul!(res::Vector{Float64}, op::LinearOperators.LinearOperator{Float64, Int64, ImplicitDifferentiation.PushforwardOperator!{ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, DifferentiationInterface.NoPushforwardExtras, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}}, DataType, Nothing, Vector{Float64}}, v::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}})
    @ LinearOperators ~/.julia/packages/LinearOperators/LoNOe/src/operations.jl:40
  [6] *(op::LinearOperators.LinearOperator{Float64, Int64, ImplicitDifferentiation.PushforwardOperator!{ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, DifferentiationInterface.NoPushforwardExtras, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}}, DataType, Nothing, Vector{Float64}}, v::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}})
    @ LinearOperators ~/.julia/packages/LinearOperators/LoNOe/src/operations.jl:47
  [7] (::ImplicitDifferentiationEnzymeExt.var"#26#30"{LinearOperators.LinearOperator{Float64, Int64, ImplicitDifferentiation.PushforwardOperator!{ImplicitDifferentiation.ConditionsXNoByproduct{typeof(conditions_components), ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}, Tuple{}, @Kwargs{}}, ADTypes.AutoEnzyme{ForwardMode{FFIABI}, false}, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}}, DifferentiationInterface.NoPushforwardExtras, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(d = 1:2, e = 3:5)}}}}, DataType, Nothing, Vector{Float64}}})(dₖx::ComponentArrays.ComponentVector{Float64, SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}, Tuple{Axis{(a = 1:2, b = 3:5, m = 6)}}})
    @ ImplicitDifferentiationEnzymeExt ~/.julia/dev/ImplicitDifferentiation/ext/ImplicitDifferentiationEnzymeExt.jl:30
.
.
.
```

**Comment by gdalle** [comment]
Awesome! I'll take a look tomorrow

**Comment by gdalle** [comment]
Out of curiosity, what are you using ImplicitDifferentiation for?

**Comment by benjaminfaber** [comment]
I'm computing the solution to a 1st-order nonlinear PDE as an optimization problem, and I would like get the sensitivity of the PDE solution wrt the PDE parameters. ImplicitDifferentiation with this fix works nicely with Enzyme to compute the Jacobian, which we can then use to compute the derivatives of other functions that are dependent on the solution of the PDE.

**Comment by gdalle** [comment]
I'm rather surprised SciML doesn't have built-in stuff for you?

**Comment by gdalle** [comment]
Very cool! I'll do my best to assist, starting with this ComponentArrays bug first thing tomorrow

**Comment by gdalle** [comment]
Okay so my investigation reveals several things at play here:

- There is indeed a ComponentArrays bug on conversion: https://github.com/jonniedie/ComponentArrays.jl/issues/266
- My Enzyme extension code is a bit weird: `dx_batch` is useless unless we do matrix-matrix multiplication with `B`, which I haven't tried yet (but should be possible thanks to https://github.com/JuliaSmoothOptimizers/LinearOperators.jl/pull/330). The right code would actually be the following:

```julia
    dc_batch = mapreduce(hcat, dx) do dₖx
        B * dₖx
    end
```

- At the moment, the right code errors because `B` encodes a pushforward of the conditions with one of the two arguments fixed. This is a case of https://github.com/EnzymeAD/Enzyme.jl/issues/1669, as shown by the following MWE. We are currently working to solve it in DifferentiationInterface with Enzyme's developer @wsmoses.

<details>
```julia
julia> using Enzyme

julia> using ComponentArrays

julia> f(x::ComponentVector, y::ComponentVector) = sum(x.a .- y.a)
f (generic function with 2 methods)

julia> x = ComponentVector(; a=[1.0, 2.0])
ComponentVector{Float64}(a = [1.0, 2.0])

julia> dx = ComponentVector(; a=[1.0, 0.0])
ComponentVector{Float64}(a = [1.0, 0.0])

julia> y = copy(x)
ComponentVector{Float64}(a = [1.0, 2.0])

julia> c = Base.Fix2(f, y)
(::Base.Fix2{typeof(f), ComponentVector{Float64, Vector{Float64}, Tuple{Axis{(a = 1:2,)}}}}) (generic function with 1 method)

julia> Enzyme.autodiff(
           Enzyme.Forward,
           Enzyme.Duplicated(c, Enzyme.make_zero(c)),
           Enzyme.Duplicated,
           Enzyme.Duplicated(x, dx),
       )
(0.0, 1.0)

julia> Enzyme.autodiff(
           Enzyme.Forward,
           c,
           Enzyme.Duplicated,
           Enzyme.Duplicated(x, dx),
       )
ERROR: Enzyme execution failed.
Mismatched activity for:   %unbox72.fca.1.0.1.insert.pn.extract.0 = phi {} addrspace(10)* [ %unbox72.fca.0.load, %L101 ], [ %getfield3, %L91 ] const val:   %getfield3 = load atomic {} addrspace(10)*, {} addrspace(10)* addrspace(11)* %getfield_addr2 unordered, align 8, !dbg !27, !tbaa !24, !alias.scope !38, !noalias !41, !nonnull !19, !dereferenceable !46, !align !47
 value=Unknown object of type Vector{Float64}
 llvalue=  %getfield3 = load atomic {} addrspace(10)*, {} addrspace(10)* addrspace(11)* %getfield_addr2 unordered, align 8, !dbg !27, !tbaa !24, !alias.scope !38, !noalias !41, !nonnull !19, !dereferenceable !46, !align !47
You may be using a constant variable as temporary storage for active memory (https://enzyme.mit.edu/julia/stable/faq/#Activity-of-temporary-storage). If not, please open an issue, and either rewrite this variable to not be conditionally active or use Enzyme.API.runtimeActivity!(true) as a workaround for now

Stacktrace:
  [1] getindex
    @ ./essentials.jl:13
  [2] getindex
    @ ./subarray.jl:323
  [3] _broadcast_getindex
    @ ./broadcast.jl:675
  [4] _getindex
    @ ./broadcast.jl:705
  [5] _broadcast_getindex
    @ ./broadcast.jl:681
  [6] getindex
    @ ./broadcast.jl:636
  [7] macro expansion
    @ ./broadcast.jl:1004
  [8] macro expansion
    @ ./simdloop.jl:77
  [9] copyto!
    @ ./broadcast.jl:1003
 [10] copyto!
    @ ./broadcast.jl:956
 [11] copy
    @ ./broadcast.jl:928
 [12] materialize
    @ ./broadcast.jl:903
 [13] f
    @ ~/Work/GitHub/Julia/ImplicitDifferentiation.jl/test/playground.jl:5

Stacktrace:
  [1] throwerr(cstr::Cstring)
    @ Enzyme.Compiler ~/.julia/packages/Enzyme/LhGXm/src/compiler.jl:1797
  [2] unalias
    @ ./abstractarray.jl:1481 [inlined]
  [3] broadcast_unalias
    @ ./broadcast.jl:977 [inlined]
  [4] preprocess
    @ ./broadcast.jl:984 [inlined]
  [5] preprocess_args
    @ ./broadcast.jl:987 [inlined]
  [6] preprocess_args
    @ ./broadcast.jl:986 [inlined]
  [7] preprocess
    @ ./broadcast.jl:983 [inlined]
  [8] copyto!
    @ ./broadcast.jl:1000 [inlined]
  [9] copyto!
    @ ./broadcast.jl:956 [inlined]
 [10] copy
    @ ./broadcast.jl:928 [inlined]
 [11] materialize
    @ ./broadcast.jl:903 [inlined]
 [12] f
    @ ~/Work/GitHub/Julia/ImplicitDifferentiation.jl/test/playground.jl:5
 [13] Fix2
    @ ./operators.jl:1135 [inlined]
 [14] Fix2
    @ ./operators.jl:0 [inlined]
 [15] fwddiffejulia_Fix2_9875_inner_1wrap
    @ ./operators.jl:0
 [16] macro expansion
    @ ~/.julia/packages/Enzyme/LhGXm/src/compiler.jl:6837 [inlined]
 [17] enzyme_call
    @ ~/.julia/packages/Enzyme/LhGXm/src/compiler.jl:6437 [inlined]
 [18] ForwardModeThunk
    @ ~/.julia/packages/Enzyme/LhGXm/src/compiler.jl:6317 [inlined]
 [19] autodiff
    @ ~/.julia/packages/Enzyme/LhGXm/src/Enzyme.jl:427 [inlined]
 [20] autodiff(mode::ForwardMode{…}, f::Base.Fix2{…}, ::Type{…}, args::Duplicated{…})
    @ Enzyme ~/.julia/packages/Enzyme/LhGXm/src/Enzyme.jl:326
 [21] top-level scope
    @ ~/Work/GitHub/Julia/ImplicitDifferentiation.jl/test/playground.jl:20
Some type information was truncated. Use `show(err)` to see complete types.
```
</details>

- I don't fully understand why your fix works, but it's probably an accident due to interactions with my weird code.

In a nutshell, I think I understand the problem, and the solution will come from a future release of ADTypes and DI that supports various activities for differentiating functions with Enzyme

**Comment by wsmoses** [comment]
@gdalle what happens if you dont use Fix1 to make a closure and just pass it as a separate arg [that'll be better for perf and compatibility anyways]

**Comment by wsmoses** [comment]
and no I don't think that's an example of the above (and that certainly should just be const)

**Comment by gdalle** [comment]
> @gdalle what happens if you dont use Fix1 to make a closure and just pass it as a separate arg [that'll be better for perf and compatibility anyways]

I know that multiple arguments are better for performance, that's like half the content of the conversations you and I have on here :rofl: But ImplicitDifferentiation is based on DifferentiationInterface to be backend-agnostic, and that's why I need to do this with a one-argument function at the moment.

> and no I don't think that's an example of the linked issue in Enzyme (and that closure certainly should just be const)

I have updated the code above to highlight another phenomenon: keeping the `Base.Fix2` closure `Const` works with `Vector` inputs but not with `ComponentVector`. That's why I think there's something fishy activity-wise, perhaps with other stuff inside the `ComponentVector` struct. What do you think? Should I open an Enzyme issue?

**Comment by wsmoses** [comment]
So multiple args doesn't just matter for perf -- but also for compatibility [as seemingly evidenced here]. Aka multi arg works, but fused would be unhappy.

Specifically in this case we have a related analysis to deduce of constant memory is stored into an active struct, which _may_ lead to incorrect results (and thus throw an error and suggest runtime activity to be on). In this case the closure adds an extra level of indirection that prevents the analysis from ensuring it runs fine (in this case the error is conservative, and not necessary)

**Comment by benjaminfaber** [comment]
Thanks @gdalle for doing all this work! Hopefully this has been helpful in improving ImplicitDifferentiation, it's looking very promising for our application.

**Comment by benjaminfaber** [comment]
Thanks @gdalle , I see this was fixed by #151 . I'll close this.

**Comment by gdalle** [comment]
It wasn't really fixed, the bug has moved ^^

**Comment by gdalle** [comment]
I haven't yet had time to investigate what the new bug is
